# Hands Classification
We have recording of hands in 3 different scenarios.  
1. Spontan- two hand moving freely.
2. Sync- two hand moving, trying to be synchronized.
3. Alone- one hand moving freely.  

Our target is to classify hands to one of the categories according to their recorded movements and position. 

In [1]:
#Libraries import
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import xgboost as xgb

In [2]:
#understanding the data
#example data:
Evyatar_spon=pd.read_csv('Datasets/Hands/Training/Evyatar Cohen/Evyatar636771052727603804Spontan.csv')

In [3]:
Evyatar_spon.head()

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0  335.9984     128556      right         2    128.63570     224.0294   
1  335.9984     128556       left         2    -93.24845     143.9423   
2  336.0150     128558      right         2    129.15550     221.9923   
3  336.0150     128558       left         2    -94.86554     148.5542   
4  336.0315     128560      right         2    129.87070     220.7944   

    Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0    -12.29588    -18.26243   -140.54450   -117.57790  ... -1.588251   
1    -37.81597   -109.09820    240.38980    -67.62959  ...  0.873458   
2    -15.00181     47.36464    -64.25022   -168.85850  ... -1.528183   
3    -39.00158    -80.99012    278.60220    -66.79356  ...  0.879703   
4    -18.38681     35.05422    -84.74242   -205.72030  ... -1.461710   

    Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0      177.5125      173.5658    -14.569630      276.8753     -33.68264   
1     -147.0455      109.9380     -5.326688     -329.8577     -34.41571   
2      177.7894      171.1952    -14.677850      276.0645     -33.15913   
3     -148.2022      113.0909     -7.375025     -326.9279     -39.97873   
4      178.3730      169.9159    -15.246240      276.1767     -30.72506   

    Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0      91.61740            0.0     1.250576              0.0  
1      93.88171            0.0     0.014736              0.0  
2      97.88171            0.0     1.069930              0.0  
3      85.98000            0.0     0.000000              0.0  
4     104.13920            0.0     0.916969              0.0  

[5 rows x 22 columns]

In [4]:
#parameters
Evyatar_spon.loc[0].index

Index(['Time', ' Frame ID', ' Hand Type', ' # hands', ' Position X',
       ' Position Y', ' Position Z', ' Velocity X', ' Velocity Y',
       ' Velocity Z', ' Pitch', ' Roll', ' Yaw', ' Wrist Pos X',
       ' Wrist Pos Y', ' Wrist Pos Z', ' Elbow pos X', ' Elbow Pos Y',
       ' Elbow Pos Z', ' Grab Strenth', ' Grab Angle', ' Pinch Strength'],
      dtype='object')

In [5]:
# frames per second
Evyatar_spon[(Evyatar_spon['Time']>336) & (Evyatar_spon['Time']<337)]

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
2    336.0150     128558      right         2    129.15550     221.9923   
3    336.0150     128558       left         2    -94.86554     148.5542   
4    336.0315     128560      right         2    129.87070     220.7944   
5    336.0315     128560       left         2    -95.16259     152.5339   
6    336.0481     128562      right         2    131.07210     219.2134   
..        ...        ...        ...       ...          ...          ...   
117  336.9594     128667       left         2    -76.96600     115.3488   
118  336.9758     128669      right         2    140.29330     167.0396   
119  336.9758     128669       left         2    -75.37846     114.4238   
120  336.9931     128671      right         2    140.24220     172.7196   
121  336.9931     128671       left         2    -73.47362     113.4763   

      Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
2     -15.001810    47.364640    -64.25022   -168.85850  ... -1.528183   
3     -39.001580   -80.990120    278.60220    -66.79356  ...  0.879703   
4     -18.386810    35.054220    -84.74242   -205.72030  ... -1.461710   
5     -39.945700   -11.372930    232.80310    -53.32079  ...  0.895820   
6     -22.055110    36.516350   -153.35370   -212.82590  ... -1.380241   
..           ...          ...          ...          ...  ...       ...   
117     8.684893    73.801640   -145.98930   -228.56920  ...  1.000582   
118   -84.622220    -5.801827    241.61630    115.49260  ... -0.478863   
119     4.754050   105.417400    -52.92049   -245.68910  ...  0.981195   
120   -83.057540    15.962420    313.52550     94.47213  ... -0.474993   
121     0.238662   117.355200    -53.55815   -275.65700  ...  0.968307   

      Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
2        177.7894     171.19520    -14.677850      276.0645    -33.159130   
3       -148.2022     113.09090     -7.375025     -326.9279    -39.978730   
4        178.3730     169.91590    -15.246240      276.1767    -30.725060   
5       -148.3343     115.92950     -9.437292     -322.0559    -44.423830   
6        179.5035     168.44820    -15.568560      276.6018    -31.228400   
..            ...           ...           ...           ...           ...   
117     -139.3870      94.56170     36.730190     -369.6905     34.979000   
118      174.9576     137.15860    -30.300330      234.9738     -9.023824   
119     -137.1051      92.73209     33.644910     -365.1103     26.101660   
120      174.3975     140.33830    -29.841730      234.8062    -16.986300   
121     -134.8678      91.34961     29.528370     -362.1878     15.724240   

      Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
2        97.88171            0.0     1.069930              0.0  
3        85.98000            0.0     0.000000              0.0  
4       104.13920            0.0     0.916969              0.0  
5        81.02193            0.0     0.000000              0.0  
6       105.96400            0.0     0.798334              0.0  
..            ...            ...          ...              ...  
117     123.28030            0.0     1.308134              0.0  
118     167.46080            0.0     0.000000              0.0  
119     121.13610            0.0     1.057895              0.0  
120     159.05110            0.0     0.000000              0.0  
121     111.30320            0.0     0.857708              0.0  

[120 rows x 22 columns]

Every ~1/60 second we have a "picture" of two hands- left and right.  
Each "picture" contains information about the hands placement, velocity and position.  
We will treat each pair of hands in a frame as one instance (single object).  
The "Alone" files will be paired with the file "HandRight.csv".  

## Data Importing

In [6]:
# data import
from os import listdir
folder=('Datasets/Hands/Training/')
Spon, Sync, Alone=list(), list(), list()
for participant in listdir(folder):
    for file in listdir(folder+participant+'/'):
        if file.endswith('Spontan.csv'):
            spon=pd.read_csv(folder+participant+'/'+file)
        elif file.endswith('Sync.csv'):
            sync=pd.read_csv(folder+participant+'/'+file)
        elif file.endswith('Alone.csv'):
            alone=pd.read_csv(folder+participant+'/'+file)
    Spon.append(spon)
    Sync.append(sync)
    Alone.append(alone)
print ('no. of df in Sync: '+str(len(Sync)))

no. of df in Sync: 9


We now have a list for every scenario.  
Test set:

In [7]:
# import test data
folder=('Datasets/Hands/Validation/')
Spon_test, Sync_test, Alone_test=list(), list(), list()
for participant in listdir(folder):
    for file in listdir(folder+participant+'/'):
        if file.endswith('Spontan.csv'):
            spon=pd.read_csv(folder+participant+'/'+file)
        elif file.endswith('Sync.csv'):
            sync=pd.read_csv(folder+participant+'/'+file)
        elif file.endswith('Alone.csv'):
            alone=pd.read_csv(folder+participant+'/'+file)
    Spon_test.append(spon)
    Sync_test.append(sync)
    Alone_test.append(alone)
print ('no. of df in Spon_test: '+str(len(Spon_test)))

no. of df in Spon_test: 3


We pair each "Alone" dataset with given right hand dataset.

In [8]:
right_hand=pd.read_csv("Datasets/Hands/HandRight.csv")
right_hand.head()

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0  128.2883      15478      right         1     15.67167     226.2064   
1  128.3048      15480      right         1     16.24119     222.9389   
2  128.3218      15482      right         1     16.72461     220.9709   
3  128.3384      15484      right         1     17.09016     219.7124   
4  128.3551      15486      right         1     17.46017     219.3652   

    Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0     20.26470     34.26665  -195.214700   -35.043020  ... -0.277800   
1     20.11781     36.16443  -180.093900    -4.052799  ... -0.277763   
2     19.94972     26.35298   -95.088190   -11.157620  ... -0.278792   
3     19.67679     19.46804   -61.755180   -21.303780  ... -0.279768   
4     19.13128     21.94964    -8.355943   -35.081790  ... -0.278629   

    Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0      46.88831      207.3548      86.23589      203.8337      49.33507   
1      47.54861      205.6568      86.42250      203.9553      47.88272   
2      48.15030      204.5182      86.35149      204.6231      45.35921   
3      48.64175      203.9989      86.16056      204.9939      44.18692   
4      48.93135      203.3777      85.59017      204.5952      42.68003   

    Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0      237.4553            0.0     0.206389              0.0  
1      238.4482            0.0     0.243050              0.0  
2      236.8544            0.0     0.281134              0.0  
3      236.0927            0.0     0.277691              0.0  
4      235.2896            0.0     0.283008              0.0  

[5 rows x 22 columns]

## Data Preprocessing

We will remove the first 7 seconds from each recording (time to get synchronized).

In [9]:
print('Example df length before: '+str(len(Alone[1])))# frames in Alone second dataframe
for scenario in (Spon,Sync,Alone):
    for df in scenario:
        df.drop(range(2*6*7),inplace=True)
print('Example df length after: '+str(len(Alone[1])))

Example df length before: 4575
Example df length after: 4491


In [10]:
# search for NaN values
c=0
for scenario in (Spon,Sync,Alone):
    for df in scenario:
        print('null in df no. '+ str(c)+': '+str(df.isna().sum().sum()))
        c+=1

null in df no. 0: 0
null in df no. 1: 0
null in df no. 2: 0
null in df no. 3: 0
null in df no. 4: 0
null in df no. 5: 0
null in df no. 6: 0
null in df no. 7: 0
null in df no. 8: 0
null in df no. 9: 0
null in df no. 10: 0
null in df no. 11: 0
null in df no. 12: 0
null in df no. 13: 0
null in df no. 14: 0
null in df no. 15: 0
null in df no. 16: 0
null in df no. 17: 0
null in df no. 18: 0
null in df no. 19: 0
null in df no. 20: 0
null in df no. 21: 0
null in df no. 22: 0
null in df no. 23: 0
null in df no. 24: 0
null in df no. 25: 0
null in df no. 26: 0


We'll reapet the process for the test data

In [11]:
print('length before reduce: '+str(len(Spon_test[0])))
# remove first 7 seconds and count NaN
c=0
for scenario in (Spon_test,Sync_test,Alone_test):
    for df in scenario:
        df.drop(range(2*6*7),inplace=True)
        print('null in df no. '+ str(c)+': '+str(df.isna().sum().sum()))
        c+=1
print('length after reduce: '+str(len(Spon_test[0])))

length before reduce: 8969
null in df no. 0: 0
null in df no. 1: 0
null in df no. 2: 0
null in df no. 3: 0
null in df no. 4: 0
null in df no. 5: 0
null in df no. 6: 0
null in df no. 7: 0
null in df no. 8: 0
length after reduce: 8885


In [12]:
# right hand data
print('length before reduce: '+str(len(right_hand)))
right_hand.drop(range(2*6*7),inplace=True)
print('null in right_hand: '+str(right_hand.isna().sum().sum()))
print('length after reduce: '+str(len(right_hand)))

length before reduce: 4812
null in right_hand: 2304
length after reduce: 4728


We have NaN values in right_hand, we need to take care of them.

In [13]:
# locate NaN values
right_hand.isna().sum()

Time                 0
 Frame ID            0
 Hand Type           0
 # hands             0
 Position X          0
 Position Y          0
 Position Z          0
 Velocity X          0
 Velocity Y          0
 Velocity Z          0
 Pitch               0
 Roll                0
 Yaw                 0
 Wrist Pos X       384
 Wrist Pos Y       384
 Wrist Pos Z       384
 Elbow pos X       384
 Elbow Pos Y       384
 Elbow Pos Z       384
 Grab Strenth        0
 Grab Angle          0
 Pinch Strength      0
dtype: int64

If the NaN values are in the same frame, we can gust drop them, losing 384 frames.  
Otherwise we will need more data.

In [14]:
print('right_hand shape with NaN: '+str(right_hand.shape))
print('right_hand shape without NaN: '+str(right_hand.dropna().shape))
print ('difference: '+str(right_hand.shape[0]-right_hand.dropna().shape[0]))

right_hand shape with NaN: (4728, 22)
right_hand shape without NaN: (4344, 22)
difference: 384


Luckily, we only have only 384 frames with missing data, we will discard them.

In [15]:
right_hand=right_hand.dropna()
right_hand.shape

(4344, 22)

### Reorganizing the data frames
We want to pair together left and right hand from the same frame.

In [16]:
# the dataframe we will demonstrate on
demo=sync.copy()
demo.head()

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
84  925.3400      60509       left         2    -121.3679     284.9276   
85  925.3400      60509      right         2     142.1444     271.0127   
86  925.3743      60511       left         2    -122.7052     287.0048   
87  925.3743      60511      right         2     141.2431     269.0698   
88  925.4076      60513       left         2    -124.1146     288.3193   

     Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
84     71.08919    -37.52761     78.14517     32.94556  ...  1.521496   
85     74.83476    -36.38242    -63.54549     54.77497  ... -1.401698   
86     72.01142    -49.97677     69.69836     35.03852  ...  1.522061   
87     76.03043    -35.50513    -88.49073     44.93065  ... -1.403901   
88     72.90015    -46.41800     35.69813     28.18376  ...  1.511062   

     Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
84     -167.1657      227.7851      67.58211     -335.6783      38.53807   
85      192.8051      217.4485      75.04996      373.2006      28.74548   
86     -168.4623      229.8258      68.45575     -334.5794      38.19341   
87      191.8715      215.4569      76.01191      377.1630      31.49922   
88     -170.4284      231.5497      69.74766     -338.1897      41.28901   

     Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
84     130.78630            0.0     0.328604              0.0  
85      82.42117            0.0     0.027603              0.0  
86     130.80850            0.0     0.332486              0.0  
87      81.69724            0.0     0.045244              0.0  
88     131.89810            0.0     0.360908              0.0  

[5 rows x 22 columns]

In [17]:
demo[' Hand Type'].value_counts()

right    2204
left     2158
Name:  Hand Type, dtype: int64

The data isn't complete. We will correct that when merging the right and left data- we will only include frames which have both hands.

Now we split the dataset and merge it horizontally
### 2 hands scenarios demo

In [18]:
demo_left=demo[demo[' Hand Type']=='left']
demo_right=demo[demo[' Hand Type']=='right']
demo_left.head()

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
84  925.3400      60509       left         2    -121.3679     284.9276   
86  925.3743      60511       left         2    -122.7052     287.0048   
88  925.4076      60513       left         2    -124.1146     288.3193   
90  925.4398      60516       left         2    -125.3973     287.9051   
92  925.4729      60518       left         2    -125.1278     284.3668   

     Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
84     71.08919    -37.52761     78.14517    32.945560  ...  1.521496   
86     72.01142    -49.97677     69.69836    35.038520  ...  1.522061   
88     72.90015    -46.41800     35.69813    28.183760  ...  1.511062   
90     73.72894    -14.55321    -53.31699    12.755390  ...  1.495221   
92     73.64545     13.44731   -139.13820    -5.644237  ...  1.492570   

     Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
84     -167.1657      227.7851      67.58211     -335.6783      38.53807   
86     -168.4623      229.8258      68.45575     -334.5794      38.19341   
88     -170.4284      231.5497      69.74766     -338.1897      41.28901   
90     -172.7207      231.9361      71.27187     -352.0691      53.22682   
92     -173.1618      229.0355      71.46091     -360.6429      60.20063   

     Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
84      130.7863            0.0     0.328604              0.0  
86      130.8085            0.0     0.332486              0.0  
88      131.8981            0.0     0.360908              0.0  
90      135.3253            0.0     0.382371              0.0  
92      138.9244            0.0     0.389056              0.0  

[5 rows x 22 columns]

In [19]:
# merge horizontally, only hands in the same frame (inner)
demo2=pd.merge(demo_left,demo_right,on=['Time',' Frame ID',' # hands'],how='inner')
demo2.head()

Time   Frame ID  Hand Type_x   # hands   Position X_x   Position Y_x  \
0  925.3400      60509         left         2      -121.3679       284.9276   
1  925.3743      60511         left         2      -122.7052       287.0048   
2  925.4076      60513         left         2      -124.1146       288.3193   
3  925.4398      60516         left         2      -125.3973       287.9051   
4  925.4729      60518         left         2      -125.1278       284.3668   

    Position Z_x   Velocity X_x   Velocity Y_x   Velocity Z_x  ...     Yaw_y  \
0       71.08919      -37.52761       78.14517      32.945560  ... -1.401698   
1       72.01142      -49.97677       69.69836      35.038520  ... -1.403901   
2       72.90015      -46.41800       35.69813      28.183760  ... -1.403654   
3       73.72894      -14.55321      -53.31699      12.755390  ... -1.405593   
4       73.64545       13.44731     -139.13820      -5.644237  ... -1.404291   

    Wrist Pos X_y   Wrist Pos Y_y   Wrist Pos Z_y   Elbow pos X_y  \
0        192.8051        217.4485        75.04996        373.2006   
1        191.8715        215.4569        76.01191        377.1630   
2        190.4933        211.9839        76.95097        379.3175   
3        190.5499        209.6013        78.05990        383.9880   
4        189.7028        206.5573        78.37885        386.9746   

    Elbow Pos Y_y   Elbow Pos Z_y   Grab Strenth_y   Grab Angle_y  \
0        28.74548        82.42117              0.0       0.027603   
1        31.49922        81.69724              0.0       0.045244   
2        31.64314        82.27242              0.0       0.077169   
3        34.35482        86.77981              0.0       0.118110   
4        35.72657        88.69302              0.0       0.124140   

    Pinch Strength_y  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0  

[5 rows x 41 columns]

We have the data of both hands under the same frame.  
Data with "_x" describes the left hand and data with "_y" describes the right hand.

In [20]:
print('demo2 length: '+str(demo2[' # hands'].size))
print('demo2 [# hands] value count: \n'+str(demo2[' # hands'].value_counts()))

demo2 length: 2158
demo2 [# hands] value count: 
2    2158
Name:  # hands, dtype: int64


We removed all single hand data.

### Alone scenario
This method will work for 2 hands data, in "Alone" datasets we need to pair the "right_hand" dataset.  
However, if we just merged every "Alone" with "right_hand" the machine will be able to classify them easily.  
To avoid such problem, we will split "right_hand" to 10 subsets numbered 0 to 9. from subsets 0 to 6 we will compose random new dataset for each "Alone" (We will choose 10 numbers between 0 to 6 and compose dataset from the corresponding subsets).  
Subsets 7 to 9 will be used in the test sets to avoid data leak.

In [21]:
# split "right_hand"
right_hand_data_list=(
    right_hand[:400],#1st split
    right_hand[400:800],#2nd split
    right_hand[800:1200],#3rd
    right_hand[1200:1600],
    right_hand[1600:2000],
    right_hand[2000:2400],
    right_hand[2400:2800] )
right_hand_data_test_list=(
    right_hand[2800:3200],#8th
    right_hand[3200:3600],
    right_hand[3600:] )
# generate constant random lists for each "Alone" 
np.random.seed(0)
right_hand_recipe=list()
for alone in Alone:
    # create right hand data numbers for each alone recording
    right_hand_recipe.append(np.random.randint(7, size=10))
#test sets
right_hand_test_recipe=list()
for alone in Alone_test:
    right_hand_test_recipe.append(np.random.randint(3, size=10))

right_hand_recipe

[array([4, 5, 0, 3, 3, 3, 1, 3, 5, 2]),
 array([4, 6, 0, 0, 4, 2, 1, 6, 6, 0]),
 array([1, 5, 1, 5, 0, 1, 4, 3, 0, 3]),
 array([5, 6, 0, 2, 3, 0, 1, 3, 5, 3]),
 array([3, 6, 0, 1, 1, 1, 0, 2, 4, 3]),
 array([6, 3, 2, 4, 2, 0, 0, 4, 5, 5]),
 array([6, 0, 4, 1, 4, 1, 2, 2, 0, 1]),
 array([1, 1, 1, 3, 6, 3, 6, 2, 3, 0]),
 array([6, 3, 5, 4, 1, 2, 4, 3, 4, 6])]

In [22]:
#create datasets by recipe
def create_dataset_from_list (datasets, recipe):
    "create dataset from the first list according to the recipe in the second list"
    df=pd.DataFrame()
    for num in recipe:
        df=df.append(datasets[num])
    return df

In [23]:
# create lists of right hand data
right_hand_train=[create_dataset_from_list(right_hand_data_list,right_hand_recipe[i]) 
                  for i in range(len(right_hand_recipe))]
# test data
right_hand_test=[create_dataset_from_list(right_hand_data_test_list,right_hand_test_recipe[i]) 
                  for i in range(len(right_hand_test_recipe))]
print('no. of right hand training datasets: '+str(len(right_hand_train)))
print('no. of right hand test datasets: '+str(len(right_hand_test)))
print('example dataset:')
right_hand_train[3]

no. of right hand training datasets: 9
no. of right hand test datasets: 3
example dataset:


Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
2084  162.3434      19404      right         1    -51.72484     192.6869   
2085  162.3599      19406      right         1    -44.04697     194.9999   
2086  162.3769      19408      right         1    -35.94160     195.9525   
2087  162.3933      19410      right         1    -28.67678     197.8434   
2088  162.4097      19412      right         1    -21.88210     199.2524   
...        ...        ...        ...       ...          ...          ...   
1679  155.5955      18626      right         1     35.62925     312.7869   
1680  155.6121      18628      right         1     34.50065     322.2206   
1681  155.6290      18630      right         1     33.62009     330.9560   
1682  155.6454      18632      right         1     32.63510     338.9619   
1683  155.6621      18634      right         1     32.06972     345.6999   

       Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
2084     45.60166    478.57050     71.90978     77.82286  ... -0.102063   
2085     47.47932    438.71950    162.01560    122.06150  ... -0.090850   
2086     49.98099    447.97540     74.39788    148.89220  ... -0.078786   
2087     52.00623    435.23320     57.74920    120.80470  ... -0.072512   
2088     54.21956    378.60280     68.50337    121.67070  ... -0.063727   
...           ...          ...          ...          ...  ...       ...   
1679     66.96964    -77.26088    545.53050    107.04390  ... -0.126339   
1680     69.21156    -61.74185    522.85270    132.21140  ... -0.125963   
1681     71.29084    -45.49222    494.93760    116.36150  ... -0.123230   
1682     73.41872    -55.72935    453.35140    119.60690  ... -0.126809   
1683     75.38303    -25.75091    407.18120    128.13660  ... -0.128730   

       Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
2084    -37.232070      192.7321      121.0896      143.8865      168.3737   
2085    -30.282170      194.8268      123.1456      146.2658      169.0971   
2086    -23.047200      195.8831      125.8824      147.1593      174.2018   
2087    -16.172820      197.0128      128.0659      147.7214      177.1716   
2088     -9.935673      197.8258      130.4503      146.9290      179.5712   
...            ...           ...           ...           ...           ...   
1679     55.407260      287.4984      136.9129      172.0479      120.1572   
1680     54.219090      296.2399      138.8883      171.0866      125.0574   
1681     53.148000      304.1263      140.6759      170.2374      129.3184   
1682     52.376140      310.9295      142.2910      170.1415      134.6808   
1683     51.939490      316.5674      143.8208      170.9357      141.9608   

       Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
2084      326.9558            0.0     0.180089              0.0  
2085      332.7809            0.0     0.219575              0.0  
2086      341.1446            0.0     0.238264              0.0  
2087      348.3457            0.0     0.273710              0.0  
2088      355.9240            0.0     0.292496              0.0  
...            ...            ...          ...              ...  
1679      320.2871            0.0     0.455127              0.0  
1680      318.5434            0.0     0.478538              0.0  
1681      316.6631            0.0     0.436282              0.0  
1682      316.3841            0.0     0.440705              0.0  
1683      318.7333            0.0     0.463626              0.0  

[4000 rows x 22 columns]

We will merge this datasets with the "Alone" datasets as described in "2 hands scenarios".

### Repeating the process on all data frames

In [24]:
#function that does the "2 hands scenarios" steps
def left_right_rehape (df):
    "pair together left and right hand from the same frame."
    left=df[df[' Hand Type']=='left'] # split
    right=df[df[' Hand Type']=='right']
    df=pd.merge(left,right,on=['Time',' Frame ID',' # hands'],how='inner') # merge horizontally
    return df 

In [25]:
#combining Alone and right_hand data
print('example shape beafore: '+str(Alone[4].shape))
Alone=[Alone[i].append(right_hand_train[i]) for i in range(len(Alone))]
Alone_test=[Alone[i].append(right_hand_train[i]) for i in range(len(Alone))]
print('example shape after: '+str(Alone[4].shape))

example shape beafore: (4847, 22)
example shape after: (8847, 22)


In [26]:
# train data
print('Example df shape before: '+str(Sync[1].shape))# frames in Alone second dataframe
Spon = [left_right_rehape(df) for df in Spon]
Sync = [left_right_rehape(df) for df in Sync]
Alone = [left_right_rehape(df) for df in Alone]
print('Example df shape after: '+str(Sync[1].shape))

Example df shape before: (8970, 22)
Example df shape after: (4485, 41)


In [27]:
# test data
print('Example df shape before: '+str(Sync_test[1].shape))# frames in Alone second dataframe
Spon_test = [left_right_rehape(df) for df in Spon_test]
Sync_test = [left_right_rehape(df) for df in Sync_test]
Alone_test = [left_right_rehape(df) for df in Alone_test]
print('Example df shape after: '+str(Sync_test[1].shape))

Example df shape before: (8881, 22)
Example df shape after: (4396, 41)
